src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -487,6 +487,48 @@ gather_bundled_assemblies_from_apk (
 	return 0;
 }
 
+int
+try_load_typemaps_from_directory (const char *path) {
+	// read the entire typemap file into a string
+	// process the string using the add_type_mapping
+	char *val = NULL;
+	monodroid_dir_t *dir;
+	monodroid_dirent_t b, *e;
+	log_info (LOG_ASSEMBLY, "load_typemaps_from_override_directory");
+	char *dir_path = utils.path_combine (path, "typemaps");
+	if (dir_path == NULL || !utils.directory_exists (dir_path)) {
+			log_warn (LOG_DEFAULT, "directory does not exist: `%s`", dir_path);
+			free (dir_path);
+			return 0;
+	}
+
+	if ((dir = utils.monodroid_opendir (dir_path)) == NULL) {
+		log_warn (LOG_DEFAULT, "could not open directory: `%s`", dir_path);
+		free (dir_path);
+		return 0;
+	}
+
+	while (readdir_r (dir, &b, &e) == 0 && e) {
+		log_warn (LOG_DEFAULT, "checking file: `%s`", e->d_name);
+		char *file_path = utils.path_combine (dir_path, e->d_name);
+		int len = androidSystem.monodroid_read_file_into_memory (file_path, &val);
+		log_info (LOG_ASSEMBLY, "read %s into memory length %d", file_path, len);
+		if (len > 0 && val != NULL) {
+			if (utils.monodroid_dirent_hasextension (e, ".mj")) {
+				add_type_mapping (&managed_to_java_maps, NULL, NULL, ((const char*)val));
+			}
+			if (utils.monodroid_dirent_hasextension (e, ".jm")) {
+				add_type_mapping (&java_to_managed_maps, NULL, NULL, ((const char*)val));
+			}
+			free (val);
```

Comment:
`add_type_mapping()` *retains* `val`; you cannot free it if you passed it to `add_type_mapping()`.

I suggest that you change the return type of `add_type_mapping()` to `bool`, and return `false` on error.

Then, you can do:

```cpp
bool added_mj = utils.monodroid_dirent_hasextension (e, ".mj") &&
    add_type_mapping (&managed_to_java_maps, NULL, NULL, ((const char*)val));
bool added_jm = utils.monodroid_dirent_hasextension (e, ".jm") &&
    add_type_mapping (&java_to_managed_maps, NULL, NULL, ((const char*)val));
if (!added_mj && !added_jm) {
    free (val);
}
```